doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,867 @@
+
+This will allow the the implementor of `apply-change` to have the freedom to optimize the operations in any order they see fit.
+
+**NOTE:** Check Patch Orderer implementation design design details in [Json_Patch_Ordering_using_YANG_Models_Design](Json_Patch_Ordering_using_YANG_Models_Design.md) document.
+
+#### 3.1.1.5 Change Applier
+This component is going to solve the problems discussed in [Stage-3 Applying list of JsonChanges in order](#stage-3-applying-list-of-jsonchanges-in-order). This component is going to help provide a mechanism for updating a JsonChange taking into account manual service restarts, and update verification. The exact implementation details of this component will not be included in this design document, but we are going to explain in details the contract for any implementation.
+
+The contract would be:
+```python
+void apply-change(JsonChange jsonChange)
+```
+
+##### 3.1.1.5.1 Apply-Change
+|aspect      |item                     |description
+|------------|-------------------------|-----------
+|inputs      |JsonChange               | It represents the changes that needs to applied to the device running config, described in [3.1.1.4.1 JsonChange](#31141-jsonchange).
+|outputs     |None                     | 
+|errors      |malformedChangeError     | Will be raised if the input JsonChange is not valid according to [3.1.1.4.1 JsonChange](#31141-jsonchange).
+|            |resourceNotFoundError    | Will be raised if running config failed to be read or in case any other external resource is not found nor available.
+|            |unprocessableRequestError| Will be raised if the change is valid, all the resources are found but when applying the change it causes an error in the system.
+|            |internalError            | Will be raised if any other error is encountered that's different than the ones listed above.
+|side-effects|updating running-config  | This operation will cause changes to the running-config according to the input JsonChange.
+|assumptions |running-config locked    | The implementor of this contract will interact with ConfigDB to updating the running-config, it is assumed the ConfigDB is locked for changes for the lifespan of the operation.
+
+Since the order of executing the operation does not matter, the implementor of this component can work on optimizing the time to run the operation. For details check [3.1.1.4 Change Applier](#3115-change-applier).
+
+**NOTE:** Check Change Applier implementation design design details in [Json_Change_Application_Design](Json_Change_Application_Design.md) document.
+
+#### 3.1.1.6 ConfigDB
+SONiC is managing configuration in a single source of truth - a redisDB instance that we refer as ConfigDB. Applications subscribe to ConfigDB and generate their running configuration correspondingly.
+
+For further details on the ConfigDB, check [SONiC Configuration Database Manual](https://github.com/Azure/SONiC/wiki/Configuration).
+
+### 3.1.2 Checkpoint
```

Comment:
I think we need the config lock here as well because nobody should be writing in the config DB when we are reading for checkpoint.

Reply:
Sure, will add it.